src/Common/DNSPTRResolverProvider.h

```diff
@@ -0,0 +1,13 @@
+#pragma once
+
+#include <memory>
+#include "DNSPTRResolver.h"
+
+namespace DB {
+    class DNSPTRResolverProvider
```

Comment:
Why do we need `DNSPTRResolverProvider`? What do you think about simply making `CARESPTRResolver` a field of `DNSResolver`?

Reply:
The idea behind the provider is to again hide the `c-ares` dependency from clients (`DNSResolver`). Without it, DNSResolver would be aware of the `c-ares` library/ depency.

Again, it's a small detail and I am fine with having it as a field. Just let me know if you prefer it this way

Reply:
Hm, actually `DB::DNSResolver` is not a client, it already hides implementation details provides interface (`DB::DNSResolver::reverseResolve`) to clients. Another layer of abstraction is overkill.

Reply:
What is the problem with caching? How does `DNSPTRResolverProvider` solve this problem? `CaresPTRResolver` can be a member of `DB::DNSResolver` and `DB::DNSResolver::reverseResolveImpl` can call `CaresPTRResolver::resolve`.

Reply:
`DB::DNSResolver::reverseResolveImpl` is static, it can't access members of `DB::DNSResolver`, right? My point was that maybe member functions could not be provided to `CachedFn`, I haven't tried it out yet. `DB::DNSPTRResolverProvider::get` is static, so it can be called from `DB::DNSResolver::reverseResolveImpl`.

The other thing is that `ares_library_init` is not thread safe. Restricting CaresPTRResolver to be a singleton that can only be retrieved via the provider allows us to have more control and safety. Ofc the same restriction could be created with `DB::DNSResolver`.

Reply:
I think it's also worth discussing `ares_library_init`. According to the docs, it's not thread safe. Looking at the impl, it indeed unconditionally updates global static variables, `ares_initialized` and `ares_init_flags`. AFAIK, the only other client of `c-ares` is `grpc`. `grpc` seems to call `ares_library_init` only under Windows, see https://github.com/grpc/grpc/blob/master/src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper.cc#L1187. Based on that, it's probably "safe" to initialize it in our side, but we should be cautious when updating `grpc` or adding new code that depends and initializes `c-ares`.

What's your opinion on this?

Reply:
Ok, so `DNSPTRResolverProvider` makes sense then, but a comment is required. 
As for `ares_library_init`, seems like we can initialize DNS resolvers (by calling `DNSResolver::instance`/`DNSPTRResolverProvider::get`/whatever) on early stage of initialization while there's only one thread (like in `Server.cpp`/`Keeper.cpp`/etc). Also we can make `ares_initialized` counter atomic, so it will be safe to call `ares_library_init` again (seems like the only reason they do not use atomic is that they want to be able to build `c-ares` without C11 or C++11 support).

Reply:
Hm... but in order to make `ares_initialized` atomic we'd have to modify `c-ares` code, right? If that's what you want, do you think it could be made in a separate PR as future work?

Reply:
Yes, that's why we have forks of most submodules we use, it will be a PR to https://github.com/ClickHouse/c-ares I think it can be done later, because no real race condition is currently possible (since grpc does not really call it)

Reply:
Cool, let's tackle it as future work then. As for the provider, I have made the CaresPTRResolver constructor public with a private token so that only the provider is able to instantiate it. As the provider was preserved, I also preserved the interface, but it can be removed if you want.

I have changed all previous usages of String to be either String or Strings. In the new implementations, I have kept it as as std::string and std::vector<std::string>. If that must be changed, should I re-declare the alias or include it from somewhere else?

Reply:
Aside from all of this, is there anything you would like to change or dive deeper?

Reply:
> I have kept it as as std::string and std::vectorstd::string. If that must be changed, should I re-declare the alias or include it from somewhere else?

It's ok to keep `std::string`. As for `std::vector<std::string>`, `Strings` looks better, but it's not mandatory (it is for longer types like `std::unordered_map<std::string, std::vector<uint64_t>>`). You can include `Core/Types.h` or re-declare `using`.

> Aside from all of this, is there anything you would like to change or dive deeper?

It mostly looks fine.